Dockerfile

```diff
@@ -0,0 +1,24 @@
+FROM alpine:3.10 AS builder
+LABEL builder=true
+
+ENV CGO_ENABLED=0
+ENV GOPATH /go
+
+RUN apk add --update -t build-deps go git mercurial libc-dev gcc libgcc
+RUN go get github.com/ericchiang/pup \
+ && cd $GOPATH/src/github.com/ericchiang/pup \
+ && go build \
+    -a \
+    -ldflags '-s -w -extldflags "-static"' \
+    -o /bin/pup
+RUN adduser -DH user
+
+FROM scratch
+
+ENTRYPOINT [ "/pup" ]
+CMD [ "--help" ]
+
+COPY --from=builder /etc/passwd /etc/passwd
```

Comment:
not needed, as you create the user again and the copy can be chowned if needed.

Reply:
See above, the user is only created in another build step, here we only apply the minimal change to "create" the user. The user itself is not really necessary, but I consider it a best practice not to run everything as `root`.

Reply:
If we use ``FROM scratch`` we can safely use the root user. There is nothing we can escape to.

Reply:
"best practice" refers more to "give future users a chance to make things right by default", so yes, `FROM scratch` minimizes the attack surface, but what happens if anybody wants to convert the image to a Windows container where we don't have an empty base image? Not that I'd do that, but I prefer to keep those minimal baselines and make people become aware of those aspects.

In fact, I don't actually care if the maintainer of `pup` thinks the same way, so I consider this PR as simple proposal which can have alternate solutions with more focus on simplicity (see your own Dockerfile you already mentioned).

Reply:
I am pretty sure that a windows variant would have at least 10 more problems cause it is windows and you can't consider everything you might maybe need in the future.

Also this would be a problem for the windows eco system when they only support half of docker.

Reply:
Yes, you're right and I fully agree. I hope you got my point, though 😉